[dtensor] expose the __create_chunk_list__ in the doc

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #144100
* #144099

as titled, this PR expose this dunder method as a public API in the doc,
so that different checkpoint implementations can leverage this protocol,
instead of exposing a separate API

cc @H-Huang @awgu @kwen2501 @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

wanchaol committed:
[dtensor] expose the __create_chunk_list__ in the doc

as titled, this PR expose this dunder method as a public API in the doc,
so that different checkpoint implementations can leverage this protocol,
instead of exposing a separate API

[ghstack-poisoned]

awgu commented on torch/distributed/tensor/_api.py:
```diff
@@ -606,6 +606,16 @@ def __create_write_items__(self, fqn: str, object: Any):
 
     def __create_chunk_list__(self):
+        """
+        Return a list of ChunkStorageMetadata, which is a data class describes the size/offset of the local shard/replica
```
nit: 
```suggestion
        Return a list of ChunkStorageMetadata, which is a dataclass that describes the size/offset of the local shard/replica
```

awgu commented on torch/distributed/tensor/_api.py:
```diff
@@ -606,6 +606,16 @@ def __create_write_items__(self, fqn: str, object: Any):
+        """
+        Return a list of ChunkStorageMetadata, which is a data class describes the size/offset of the local shard/replica
+        on the current rank. For DTensor, each rank will have a single local shard/replica, so the returned list usually
+        only have one element.
```
```suggestion
        only has one element.
```

wanchaol committed:
Update on "[dtensor] expose the __create_chunk_list__ in the doc"


as titled, this PR expose this dunder method as a public API in the doc,
so that different checkpoint implementations can leverage this protocol,
instead of exposing a separate API

cc H-Huang awgu kwen2501 fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

wanchaol commented:
@pytorchbot merge

malfet commented on torch/distributed/tensor/_api.py:
```diff
@@ -606,6 +606,16 @@ def __create_write_items__(self, fqn: str, object: Any):
             raise RuntimeError("Unsupported tensor type!")
 
     def __create_chunk_list__(self):
```
Hmm, dunder functions usually has some special meaning in Python runtime. What is it decorated like that?

wanchaol replied:
I am wondering what's the special meaning refers to? iirc some dunder methods do operator overload, while others retrieves object information. This method is mainly retrieves the chunk storage metadata for distributed checkpoint purpose.
